### ISSUE TITLE
shoulda-matchers >4.1.1 is incompatible with the active_model_attributes gem

### ISSUE THREAD

**Issue by shevaun** [thoughtbot/shoulda-matchers#1288]
This is mostly an FYI, since I'm not sure of the best way to resolve this, but:

[These two methods](https://github.com/thoughtbot/shoulda-matchers/blob/v4.3.0/lib/shoulda/matchers/rails_shim.rb#L141-L163) in `rails_shim.rb` use `model.respond_to?(:type_for_attribute)` to determine whether the class under test is an ActiveRecord or ActiveModel (the former returns true and the latter returns false).

However, the [`active_model_attributes`](https://github.com/Azdaroth/active_model_attributes) gem also [defines the method](https://github.com/Azdaroth/active_model_attributes/pull/11)  `type_for_attribute`, so using `validate_presence_of` on objects that include the `ActiveModelAttribute` module blow up with an `NoMethodError: undefined method 'columns' for X` error, when it hits [line 145](https://github.com/thoughtbot/shoulda-matchers/blob/v4.3.0/lib/shoulda/matchers/rails_shim.rb#L145).

**Comment by mcmire** [comment]
Hey @shevaun. Yeah, looking at this again, we probably need to change `if model.respond_to?(:type_for_attribute)` to `if model.respond_to?(:columns) && model.respond_to?(:type_for_attribute)`. That would cause NotImplementedError to get raised, which would get picked up at line 157 and then [line 159](https://github.com/thoughtbot/shoulda-matchers/blob/v4.3.0/lib/shoulda/matchers/rails_shim.rb#L159) would get returned. That would fix this issue.

ActiveModel::Attributes is supported in Rails 6, so I feel like we could probably test against that to indirectly confirm that shoulda-matchers works with this gem.

**Comment by shevaun** [comment]
@mcmire Cool thanks, and yep that would fix it - we've written a very similar monkey patch to get it working in our Rails app.
I'm happy to create a PR with this change if you like? Just let me know.

**Comment by mcmire** [comment]
@shevaun Yes that would be great!

**Comment by shevaun** [comment]
@mcmire I started doing this today, but as I was looking at `ActiveModel::Attributes` in Rails 6 I realised that it implements the `attribute_types` method, so it will never reach the problematic `model.respond_to?(:type_for_attribute)` code. 

So I figured a better solution might be for the ActiveModelAttributes gem to implement `attribute_types`, thus bringing its interface closer to ActiveModel::Attributes. I've created a [PR](https://github.com/Azdaroth/active_model_attributes/pull/15) on their repo, so will see what happens with that.
